### PR TITLE
Show text labels in block toolbars when option is set.

### PR DESCRIPTION
**Pull request by tellthemachines**
## Description

Follows on from #24304 and addresses most of remaining work for #15311 .

## How has this been tested?

Tested across browsers.

## Screenshots 

Top toolbars are working on large and small  breakpoints (small with horizontal scrollbar)

Screen Shot 2020-10-15 at 1 51 15 pm

Screen Shot 2020-10-15 at 1 52 39 pm

Inline toolbar gets a bit long, not sure how we can improve it:

Screen Shot 2020-10-15 at 1 52 08 pm

## Types of changes

New feature (non-breaking change which adds functionality)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by mapk** [comment]
This looks super difficult! Thanks for taking it on, @tellthemachines! 🙏 

Here's a quick mockup I put together to work toward in this PR.

- I renamed a few items trying to condense them more.
- I removed the "Change block type or style" in hopes that just using the block name would be sufficient.
- I put all formatting options into a dropdown. With the text label option, the toolbar gets too long when Bold, Italic, and Link are surfaced on the top level.
- I like the idea of stacking "move up" and "move down" so let's give it a whirl. 

Paragraph toolbar 2

**Comment by ZebulanStanphill** [comment]
@mapk 
>I removed the "Change block type or style" in hopes that just using the block name would be sufficient.

I think that kind of defeats the whole point of this text label setting. Perhaps "Transform block" would be a better short label?

**Comment by mapk** [comment]
>Perhaps "Transform block" would be a better short label?

I thought about that, but I also felt the name of the block was equally important. I wonder if we'd want to work in a slight redesign in this mode. Something like this?

Paragraph toolbar 3

**Comment by ZebulanStanphill** [comment]
It's important to remember that the labels will be read out by screen readers, and I'm not sure a button labeled "paragraph transform block" will make much sense to someone using a screen reader.

I think this goes back to an issue I've pointed out before, where the block toolbar's transform button is serving dual function as an indicator of the current block type. This is already questionable in my opinion when icons are used, but as can be seen here, it becomes a lot more prominent when the buttons use text labels.

I think the current block type label needs to be split apart from the transform button. Semantically, it must not be inside the `<button>`.

What if we showed the current block type label in a thin strip above all the block toolbar buttons? That would give us plenty of room to display the full block title (along with the icon), without taking away (horizontal) space from any actual toolbar buttons. If we did this on both text and icon modes, we could change the transform button to use a dedicated "transform" icon, since it would no longer be acting as the block indicator.

**Comment by etoledom** [comment]
📳✅  Tested on mobile and everything seems to be working as expected

**Comment by tellthemachines** [comment]
I'm finally getting back to this, and reading through all the comments. Thanks everyone for the feedback so far!

> It's important to remember that the labels will be read out by screen readers, and I'm not sure a button labeled "paragraph transform block" will make much sense to someone using a screen reader.

Good point, and having the block name here may not be very helpful to a screen reader user because, in order to get to the block toolbar, they have to select the block itself first, so will already have had the block name announced at that step. It might be more useful if they access the transform button from a form controls menu, but then it will only will show up if a particular block is selected, meaning the user will have interacted with the block just before invoking the menu. 

The block name is important for sighted users though. 

> What if we showed the current block type label in a thin strip above all the block toolbar buttons?

Would this be only when the text labels are enabled, or with the icons too?

**Comment by ZebulanStanphill** [comment]
@tellthemachines
> > What if we showed the current block type label in a thin strip above all the block toolbar buttons?
> 
> Would this be only when the text labels are enabled, or with the icons too?

I'd say enable it for both. I think there are already a lot of cases where it's really hard to tell what block you're editing based on the icon: Columns versus Column, Buttons versus Button, the Query and Query Loop blocks, and so on. And that's not even taking into account plugin-added blocks such as a plugin Columns block versus a core Columns block, and etc. The inspector should not be relied on for this info because it's supposed to be for secondary/uncommonly-used options, not primary info like the block title. Additionally, mobile users will not be able to have the inspector open all the time due to limited screen space, so ideally, we should make info like the block title clearly visible without having it open.

Additionally, if we displayed the block title on its own little strip, we could move the block icon there as well, and we could give the transform button a dedicated icon, which would be an accessibility improvement.

**Comment by tellthemachines** [comment]
I removed the block name from the switcher button for now, and shortened some of the labels as per @mapk's suggestion:

Screen Shot 2021-01-06 at 3 47 06 pm

Screen Shot 2021-01-06 at 3 46 58 pm

I'm not sure where to place the block name though, considering that the parent block button pops up just above the toolbar, which would be its most natural location:

Screen Shot 2021-01-06 at 3 47 26 pm

As for moving the "Bold", "Italic" etc. into a "Formatting" dropdown, it won't be too straightforward because the same formatting toolbar is shared between the block toolbar and the little one that pops up in captions. It feels a bit weird to turn that one into a dropdown and hide all the options. Even for the block toolbar I'm not 100% sure about it, as those are pretty frequently used options.

**Comment by jasmussen** [comment]
This is the challenge we knew we'd run into. Thank you for your patience with this.

The shortening of labels has already helped a great deal. I'd think we could go further:

- More formatting → More
- More options → Options
- Transform block → Transform
- Transform 3 blocks → Transform
- Select parent (Cover) → Select Cover

> I removed the block name from the switcher button for now

I'm seeing this:

Screenshot 2021-01-06 at 09 41 01

Am I seeing an old version of the branch? I'm getting the sense from the conversation that this is supposed to be two buttons now?

- In any case, if it's one button, I'd say just go with "Paragraph". Have it be the button that includes transformations, styles, and pop out the "select parent" button just as usual.
- If it's two buttons, then I'd go with "Paragraph" and "Transform". Paragraph pops out the "select parent" button, and "Transform" does the transformations and styles. 

I'd lean towards one button, because it's weird that what's meant to be a visual change also affects the number of buttons. If I'm misunderstanding anything here, I trust you'll let me know.

High level, I would encourage all to keep two things in mind:

- This is a priority pull request. Even small improvements can have big improvements to those looking for this option. And getting an initial basic version of this PR in lets that happen sooner, and does not preclude additional PRs  based on what we learn.
- A button label is not the same as "title text". Button labels should indicate the action taken, and you can still have a title attribute that provides further description. In that vein, you can easily have a "Paragraph" button label, and keep the full title tag that says "Change block type or style (Paragraph)".

The MacOS Finder has a text-only option that's a good example of this:

Screenshot 2021-01-06 at 09 47 29

"View" is the action, and when you hover it, you get additional descriptive text.

Finally, if we set the font size to 12px (`$helptext-font-size`) for buttons when this option is enabled, we free up a lot of space while still being legible.

**Comment by jasmussen** [comment]
This is a good example of the idea of having dedicated button labels rather than just using the `title` text for button labels:

Screenshot 2021-01-06 at 09 40 53

For text labels I'd expect "Unordered list" or just "Unordered", and the title text could then remain "Convert to unordered list".

**Comment by tellthemachines** [comment]
Thanks for helping out with this @jasmussen !

> Am I seeing an old version of the branch? I'm getting the sense from the conversation that this is supposed to be two buttons now?

No, that was me forgetting to push the changes, sorry 😅 

> I'd lean towards one button, because it's weird that what's meant to be a visual change also affects the number of buttons.

Yup, we should have just one button. What @ZebulanStanphill was suggesting and we were discussing is having a label with the block title separate from that button. The problem is there doesn't seem to be a good place to put that label.

> A button label is not the same as "title text". Button labels should indicate the action taken, and you can still have a title attribute that provides further description.

The problem in this case is that, to make this work, we're displaying the button `aria-label` text (which is used as title text, for the tooltip) in a pseudo-element, and hiding the SVG icon. So label and title text are the same. We adopted this strategy for the previous top toolbar work to avoid making changes to the button component as well as all the places it's used in, based on the assumption that the `aria-label` content should always be the same as the actual button label. That's how we use it in the icon buttons anyway: a screen-reader interprets the `aria-label` as the actual label. 

I reckon in 99% of the cases this strategy serves us well, but the Transform button is a bit of an edge case because it does several things at once: shows the block title, opens the transforms menu and provides access to the parent block button.

If we were to add some descriptive text in addition to the `aria-label`, our best bet accessibility-wise would probably be `aria-describedby` on the button, pointing to a span with the description. The `title` attribute is known to have poor/inconsistent screen reader support, so it's best to avoid it (that's probably why we're using `aria-label` for the tooltips instead of `title`).

> For text labels I'd expect "Unordered list" or just "Unordered", and the title text could then remain "Convert to unordered list".

In this case it would actually be better if we were using the plain button label for "Unordered" and and then setting the `aria-label` to "Convert to unordered list", because if we use a combination of `aria-label` and `aria-describedby`, screen readers will end up reading something like "Unordered list convert to unordered list" which sounds terrible 😬 

The other consideration regarding tooltips is that they don't work on touchscreens or with keyboard-only navigation, so we shouldn't really be relying on them for any essential information.

I'm not at all sure what the best approach to this should be. Any thoughts appreciated!

**Comment by tellthemachines** [comment]
I've updated the text on most of the buttons, but left the "Transform" in place as I think just the block title will, in most cases, be even less informative.

Screen Shot 2021-01-07 at 2 31 50 pm

There's still a teensy bit of horizontal scroll on some toolbars, at the smallest breakpoint before the toolbar moves to the top, but we can't really fix that by shortening the labels, because we can't know how they'll look in different languages. It's better than it was, at any rate. I think the biggest remaining issue is how to solve Transform button/block title in the way that's least confusing for everyone 😅

**Comment by jasmussen** [comment]
Thank you for your patience in explaining this to me again — I know there's a lot of context that goes by in the archives that is easily missed. Makes total sense. 

> If we were to add some descriptive text in addition to the aria-label, our best bet accessibility-wise would probably be aria-describedby on the button, pointing to a span with the description. The title attribute is known to have poor/inconsistent screen reader support, so it's best to avoid it (that's probably why we're using aria-label for the tooltips instead of title).

It's a pleasure to watch you work. I'm going to trust and defer totally to you on the implementation side. 

> In this case it would actually be better if we were using the plain button label for "Unordered" and and then setting the aria-label to "Convert to unordered list"

This _sounds_ correct to me: an actionable button label, with optional additional help in tooltip or expanded describedby help.

> There's still a teensy bit of horizontal scroll on some toolbars, at the smallest breakpoint before the toolbar moves to the top, but we can't really fix that by shortening the labels, because we can't know how they'll look in different languages. 

Yep, this is why icon buttons exist in the first place, and all we can do is try and make it as good as possible within those constraints. Overflow menus, which put items in dropdowns that even fly-out, might be a thing to explore, but should not preclude this PR from happening first. Did you try a 12px font size?

> I think the biggest remaining issue is how to solve Transform button/block title in the way that's least confusing for everyone 😅

and ...

> I've updated the text on most of the buttons, but left the "Transform" in place as I think just the block title will, in most cases, be even less informative.

I don't think these two are easily reconcilable. The transformations and style changes are properties of the block itself, that's why they are in a menu related to the block itself. In that vein, I still think the best approach, especially in order to unblock this PR from happening, is to keep the block title as the label, and have title or describedby text help explain what actions you can take on the block. 

There are separate discussions on how to best establish the semantic groupings in the block toolbar in https://github.com/WordPress/gutenberg/issues/25983, and it's fine to discuss any shortcomings of the block title/indicator and how transforms are part of that. And hopefully such conversations can lead to improvements for both the cases of icon only and label only configurations in a consistent way. But in the mean time, fragmenting the experience doesn't seem like a good idea to me.

**Comment by tellthemachines** [comment]
@jasmussen thanks for talking this through with me! 

I have updated the block switcher to use the block name as label, and added an optional `aria-describedby` attribute to the Button component for use in these cases. When it exists, the description text also displays in the tooltip. 

I also reduced the button label font size to 12 in the block toolbar, so now there's less chance of horizontal scroll occurring 😅 

How does this seem for a first iteration?

File: packages/block-editor/src/components/block-toolbar/style.scss
Comment (jasmussen) [inline packages/block-editor/src/components/block-toolbar/style.scss:122]: Can you use `$helptext-font-size` here instead of rem units? We use this same variable (which returns 12px) for form widget help text.

**Comment by jasmussen** [comment]
This has been quite a journey. But I think it makes for a great first version:

Screenshot 2021-01-11 at 08 29 13

- It shows text labels for everything, accomplishing the primary goal.
- It mostly avoids cropped toolbars on the desktop breakpoint.
- Along the way you elegantly managed to tighten up and refine text labels and help text.

It is reminiscent of how you in the Apple Finder can have text only labels:

Screenshot 2021-01-11 at 08 41 30

Additionally, and I think this is important for people following along: this PR doesn't preclude further enhancements. Whatever shortcomings have been discussed, these can still be addressed, while still shipping benefits today.

---

Speaking of things that are worth revisiting in the future, and very probably some of these aren't specific to this PR... 

Screenshot 2021-01-11 at 08 46 58

Worth exploring a better way to collapse the tools on the left. The "Options" menu on the right is already an overflow button and doesn't get a hamburger icon because of it.

Screenshot 2021-01-11 at 08 53 18

While the select and edit tools technically invoke modes, they are still tools, just like how the Photoshop Marquee tool invokes a mode. I suspect the two labels are related, because if you collapse tools into tools it's a bit of a double whammy, so worth looking at together.

Screenshot 2021-01-11 at 08 52 42

Text labels aren't really centered. 

Screenshot 2021-01-11 at 08 52 47

The padding left and right for "Align" and the inline tools is different from that of Paragraph and Options. Possibly related to #24898.

Screenshot 2021-01-11 at 08 53 48

Probably related to the padding, the focus styles are slightly tight for some buttons. Also probably a result of the slightly hacky grid system for the toolbar.

The tooltips aren't centered:

Screenshot 2021-01-11 at 08 54 01

Screenshot 2021-01-11 at 08 54 08

I wonder if that's related to #27993? I know that the misaligned dropdown below is a result of that:

Screenshot 2021-01-11 at 09 02 56

When every button is a text label, should there be a tooltip that says the same?

Screenshot 2021-01-11 at 08 54 13

I don't think all of the items above have to be addressed as part of this PR, it's gone on for a while. But they may be worth extracting into a separate ticket so they can be looked at by someone with bandwidth, separately. For now, thanks for your work, @tellthemachines!

File: packages/block-editor/src/components/rich-text/format-toolbar-container.js
Comment (talldan) [inline packages/block-editor/src/components/rich-text/format-toolbar-container.js:22]: What does this do?

File: packages/block-editor/src/components/alignment-toolbar/index.js
Comment (talldan) [inline packages/block-editor/src/components/alignment-toolbar/index.js:41]: I feel a bit nervous about changing the label to something so succinct, are we confident it's informative enough for screen reader users?

This one in particular because it's using the same label as block alignment.

File: packages/block-editor/src/components/block-parent-selector/index.js
Comment (talldan) [inline packages/block-editor/src/components/block-parent-selector/index.js:63]: I don't think this text is clear enough. If you have a group block within a group block, the text 'Select Group' isn't helpful.

'Select Parent' would probably be better if it needs to be shortened.

**Review by talldan — COMMENTED** [review]
It looks really good compared to what I thought it'd look like, great job!

I think there's only small adjustments required now.

I've held off approving, as I don't it'd be right to merge the 'Select parent' button as it is.

Also noticed that some buttons show the exact same text in a tooltip, but that's minor:
Screenshot 2021-01-11 at 6 10 38 pm

File: packages/block-editor/src/components/block-parent-selector/index.js
Comment (talldan) [reply packages/block-editor/src/components/block-parent-selector/index.js:63]: Looking at the UI, I don't think shortening is required as much for this button. It's on its own line so doesn't cause the toolbar to get much longer:
Screenshot 2021-01-11 at 6 17 30 pm

Some block names are also quite long compared to the word 'parent' (e.g. Navigation Link), so I don't think this change has a huge impact.

Comment (draganescu) [reply packages/block-editor/src/components/block-parent-selector/index.js:63]: I also agree with @talldan about not shortening the select parent button's label. Being the only one in its row, it can be left as is.

**Comment by draganescu** [comment]
I love this! I have been using text only labels in OSX and also pretty much everywhere I could, including in Gutenberg as they are now. This PR makes them even better, as I am one of those for whom icons add a useless cognitive load. I expected the look of this to be much worse, so @tellthemachines and @jasmussen you both are need a 🏆  

Lovely :)

**Comment by jasmussen** [comment]
I think there are appropriate places for icons, and I usually refer to bold and italic as great examples. But I appreciate that this is becoming an option people can enable! Also, the medal should wholly go to @tellthemachines!

File: packages/block-editor/src/components/rich-text/format-toolbar-container.js
Comment (tellthemachines) [reply packages/block-editor/src/components/rich-text/format-toolbar-container.js:22]: It makes the format toolbar use a popover slot located inside the interface skeleton (same as the block toolbar uses), which is important because the `show-icon-labels` class is added to the skeleton container. By default the toolbar would use the popover slot outside it, so it wouldn't get the relevant styles.

File: packages/block-editor/src/components/alignment-toolbar/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/alignment-toolbar/index.js:41]: Hmm, might be worth adding the full text as a tooltip here.

File: packages/block-editor/src/components/block-parent-selector/index.js
Comment (tellthemachines) [reply packages/block-editor/src/components/block-parent-selector/index.js:63]: That makes sense!

File: packages/block-editor/src/components/block-toolbar/style.scss
Comment (tellthemachines) [reply packages/block-editor/src/components/block-toolbar/style.scss:122]: Sure! I didn't find that variable (it's not with the other font sizes at the top of the file), so thought I'd sneakily introduce a responsive unit instead 😄

**Comment by tellthemachines** [comment]
I'm loving how everyone assumed this would inevitably look like the dog's breakfast 😂 

Review feedback addressed, except for:

> The padding left and right for "Align" and the inline tools is different from that of Paragraph and Options.

@jasmussen this is how padding is set when the icons show: all buttons have 6px each side, except for the first and last which have 12px. Should I reduce these to 6px when text labels are visible?

> The tooltips aren't centered
> I wonder if that's related to #27993?

Yup, longstanding popover issue 😬 

> When every button is a text label, should there be a tooltip that says the same?

I agree we should remove the tooltip where it's redundant, but in order to do that we'll have to make `block-editor` aware of the text label setting, which currently lives inside `edit-post`. I'd rather leave that for a separate PR, as I'm still thinking about the best way to go about it. With the site editor, widgets and other editors coming up, it might make sense to move the setting out of `edit-post` and make it global.

**Comment by tellthemachines** [comment]
I reverted changes to the alignment labels as I was running into an issue where passing `toggleProps` to `ToolbarGroup` was causing the whole toolbar keyboard navigation to fail. As I'm still not sure of the root cause of this, I'd rather address it as a separate issue. For now, I settled for using "Align text" and "Align block" to differentiate between them (Change vertical alignment is unchanged).

**Comment by jasmussen** [comment]
> this is how padding is set when the icons show: all buttons have 6px each side, except for the first and last which have 12px. Should I reduce these to 6px when text labels are visible?

I think there's some gnarly hacks going on here, which I hope to clean up when I get a chance to, as part of #24898. Specifically there's some child rules that aim to optically space icons, which might just mess with things on the text labels. Probably best to leave it alone for now.

> I'd rather leave that for a separate PR

👍 👍

**Comment by tellthemachines** [comment]
Ok peeps, this one's ready for a final review. Many thanks to @diegohaz for fixing the issue with the alignment labels ⭐

**Comment by draganescu** [comment]
I found this little stuffed text (Edge browser):

edge-caption-text-fit

**Comment by tellthemachines** [comment]
@draganescu I can't reproduce that locally:

Screen Shot 2021-01-15 at 9 29 38 am

That string was shortened [here](https://github.com/WordPress/gutenberg/blob/61d7d95259b788f705c9b96bde609447a12db893/packages/block-editor/src/components/rich-text/format-toolbar/index.js#L45). Could it be an issue with your local files, or maybe aggressive browser caching?

**Review by draganescu — APPROVED** [review]
The code looks great IMO and my testing found no issues whatsoever. I did not test if this affects the screen reader experience negatively.

File: packages/block-editor/src/components/alignment-toolbar/index.js
Comment (hypest) [reply packages/block-editor/src/components/alignment-toolbar/index.js:41]: Came here to say something similar to Dan. As part of fixing a regression after this PR got merged, it seems that we [rely on now single-word label](https://github.com/WordPress/gutenberg/pull/28286/files#diff-4749eb3277e22380531b254004b270c0d603cc201321292f13572abf7509a6f0R27) to find the button in E2E tests and tap it on native mobile. My small concern is whether that single word label is unique enough for confident search of it. I don't think it's an issue atm, but wanted to share the concern.

Comment (tellthemachines) [reply packages/block-editor/src/components/alignment-toolbar/index.js:41]: We can always resort to a combination of classnames, or other attributes, and labels to find the item. E2e tests shouldn't determine UI design 😅

File: packages/components/src/button/index.js
Comment (talldan) [inline packages/components/src/button/index.js:109]: This `uniqueId` will change on every render, which is unnecessary.

I think it should use the `useInstanceId` hook.

File: packages/components/src/button/index.js
Comment (youknowriad) [inline packages/components/src/button/index.js:144]: There's some places where we assume buttons only render a single "parent" element. In the block toolbar for instance.

So when you render two buttons in the block toolbar that have descriptions, the spacing is not evenly distributed. See https://github.com/WordPress/gutenberg/pull/39605#issuecomment-1073644254

Is it possible to move the visually hidden text inside the "button" or "a" directly? Would that be valid HTML? Is there any other alternative?

Comment (ZebulanStanphill) [reply packages/components/src/button/index.js:144]: It would probably be valid HTML, but that doesn't mean it would be valid a11y. Putting it inside the `<button>`/`<a>` would make the whole thing be considered part of the button label. I don't think that would be correct/desirable, but I could be mistaken.

Comment (tellthemachines) [reply packages/components/src/button/index.js:144]: Yeah, the description shouldn't be inside the element being described. 

Potentially we could find another workaround by adding in actual button text as we did with the global inserter.

But the visually hidden elements are absolutely positioned; I wouldn't expect them to occupy any space at all in the canvas. Could that spacing be caused by some other issue?

Comment (youknowriad) [reply packages/components/src/button/index.js:144]: > Could that spacing be caused by some other issue?

We have some styles targetting last-child and first-child CSS pseudo elements causing these issues. So these addition conflicts with these styles. I'm considering exploring changes to these styles but with no guarantee of success.

Comment (tellthemachines) [reply packages/components/src/button/index.js:144]: Oh I see, the problem is not `VisuallyHidden` but the pseudo-element we use to render the text label. In that case, we could definitely explore a similar approach to the global inserter: render actual text in the button instead of the pseudo-element.

**Pull request merged**
